File: code/ex283.java
class Solution {
    public void moveZeroes(int[] nums) {
        // Get numbers of zeros in the array
        int zeros = 0;
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] == 0)   zeros++;
        }
        
        // Rearrange the array
        int counter = 0;
        int n = 0, i = 0;
        
        while (counter < nums.length - zeros) {
            if (nums[i] != 0) {
                nums[n] = nums[i];
                
                i++;    n++;    counter++;
            }
            else
                i++;
        }
        
        // Add zeros to the end
        for (int j = nums.length - zeros; j < nums.length; j++) nums[j] = 0;
        
    }
}

File: code/ex283.cpp
class Solution {
public:
    void moveZeroes(vector<int>& nums) {
        int n = nums.size();
        int lo = 0, hi = 0;
        while (hi < n) {
            while (hi < n && nums[hi] == 0)    hi++;
            if (hi < n) {
                swap(nums[lo++], nums[hi++]);
            }
        }
    }
};